agent/engine/docker_image_manager.go

```diff
@@ -388,70 +388,82 @@ func (imageManager *dockerImageManager) getNonECSContainerIDs(ctx context.Contex
-	}
 	// we want to sort images with size ascending
-	sort.Slice(imageWithSizeList, func(i, j int) bool {
-		return imageWithSizeList[i].Size < imageWithSizeList[j].Size
+	sort.Slice(nonECSImages, func(i, j int) bool {
+		return nonECSImages[i].Size < nonECSImages[j].Size
 	})
 
 	// we will remove the remaining nonECSImages in each performPeriodicImageCleanup call()
 	var numImagesAlreadyDeleted = 0
-	for _, kv := range imageWithSizeList {
+	for _, image := range nonECSImages {
 		if numImagesAlreadyDeleted == nonECSImagesNumToDelete {
 			break
 		}
-		seelog.Infof("Removing non-ECS Image: %s", kv.ImageName)
-		err := imageManager.client.RemoveImage(ctx, kv.ImageName, dockerclient.RemoveImageTimeout)
+		seelog.Debugf("Removing non-ECS Image: %s (Tags: %s)", image.ImageID, image.RepoTags)
+		err := imageManager.client.RemoveImage(ctx, image.ImageID, dockerclient.RemoveImageTimeout)
 		if err != nil {
-			seelog.Errorf("Error removing Image %s - %v", kv.ImageName, err)
+			seelog.Errorf("Error removing Image %s (Tags: %s) - %v", image.ImageID, image.RepoTags, err)
 			continue
 		} else {
-			seelog.Infof("Image removed: %s", kv.ImageName)
+			seelog.Infof("Image removed: %s (Tags: %s)", image.ImageID, image.RepoTags)
 			numImagesAlreadyDeleted++
 		}
 	}
 }
 
-func (imageManager *dockerImageManager) getNonECSImageNames(ctx context.Context) []string {
-	response := imageManager.client.ListImages(ctx, dockerclient.ListImagesTimeout)
-	var allImagesNames []string
-	for _, name := range response.RepoTags {
-		allImagesNames = append(allImagesNames, name)
+// getNonECSImages returns type ImageWithSizeID with all fields populated.
+func (imageManager *dockerImageManager) getNonECSImages(ctx context.Context) []ImageWithSizeID {
+	r := imageManager.client.ListImages(ctx, dockerclient.ListImagesTimeout)
+	var allImages []ImageWithSizeID
+	// inspect all images
+	for _, imageID := range r.ImageIDs {
+		resp, err := imageManager.client.InspectImage(imageID)
+		if err != nil {
+			seelog.Errorf("Error inspecting non-ECS image: (ImageID: %s), %s", imageID, err)
+			continue
+		}
+		allImages = append(allImages,
+			ImageWithSizeID{
+				ImageID:  imageID,
+				Size:     resp.Size,
+				RepoTags: resp.RepoTags,
+			})
 	}
-	var ecsImageNames []string
+
+	// get all 'ecs' image IDs
+	var ecsImageIDs []string
 	for _, imageState := range imageManager.getAllImageStates() {
-		for _, imageName := range imageState.Image.Names {
-			ecsImageNames = append(ecsImageNames, imageName)
+		ecsImageIDs = append(ecsImageIDs, imageState.Image.ImageID)
+	}
+
+	// exclude 'ecs' image IDs and image IDs with an explicitly excluded tag
+	var nonECSImages []ImageWithSizeID
+	for _, image := range allImages {
+		if !isInExclusionList(image.ImageID, ecsImageIDs) {
+			var exclude bool
+			for _, repoTag := range image.RepoTags {
+				if isInExclusionList(repoTag, imageManager.imageCleanupExclusionList) {
```

Comment:
this can be refactored as - 
```
if !isInExclusionList(repoTag, imageManager.imageCleanupExclusionList) {
  nonECSImages = append(nonECSImages, image)
}
```
avoids the use of local var

Reply:
The above if-statement would add the image ID to "nonECSImages" if _any_ tag is not in the exclusion list. The local variable is necessary because we want to skip Image IDs that have any tag in the exclusion list.

ie, an ImageID with 1 out of 10 tags in the exclusion list should be excluded.